File: Java/cp/ReverseString.java

/* Program to reverse a string */

import java.io.*;
import java.util.*;
 
// Main class

class Program {
    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);

        // Accepting input from user

        System.out.println("Enter string");
        String str = sc.nextLine();

        // revoking the function reverse

        reverse(str);
        
    }

    // Method for reversing the string

    static void reverse(String s){

        // Reversing the string by extracting each element from the end of the string and storing in a variable result

        String result = "";
        for (int i = s.length() - 1; i >= 0; i--)
            result += s.charAt(i);

        // Printing the reversed string

        System.out.println(result);
       
    }

}

/* Time Complexity is O(n)
   Space Complexity is O(n)
   
   Input :
   Hello World
   
   Output :
   dlroW olleH
*/

File: Kotlin/cp/ReverseString.kt
/* Program to reverse a given sentence using a recursive loop in Kotlin
We will create a recursive function and operate it till the end and store the reverse order.
*/


import java.util.Scanner       

fun main(args: Array<String>) {

   val reader = Scanner(System.`in`)       
   print("Enter any sentence: ")  
   var sentence :String = reader.nextString()  
   
    val reversed = reverse(sentence)
    println("The reversed sentence is: $reversed")
}

fun reverse(sentence: String): String {
    if (sentence.isEmpty())
        return sentence

    return reverse(sentence.substring(1)) + sentence[0]
}

/*
Sample Test case:
Enter any sentence: Hello Kotlin 
The reversed sentence is : niltoK olleH
*/

/*
Time Complexity : O(n)
Auxiliary Space : O(1)
*/

